### PR TITLE
[Tabular] Quantile aggregation model (updated version)

### PR DESCRIPTION
**Pull request by taesup-aws**
*Description of changes:
1.  Add a neural network based quantile aggregation (ensemble) model. (cc @jwmueller)
- `TabularQuantileAggregatorTorchModel`: PyTorch based, aggregation (ensemble) model. 
-  This is wrapped by `LinearAggregatorModel` and is only activated when `self.bagged_mode` is `True`.
2.  Also minor fix for quantile regression models.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tabular/src/autogluon/tabular/predictor/predictor.py [hunk 1]
```diff
@@ -889,7 +895,7 @@ def _calibrate_model(self, model_name: str = None, lr: float = 0.01, max_iter: i
         if model_name is None:
             model_name = self._trainer.get_model_best()
 
-        if self._trainer.bagged_mode:
+        if self._trainer.bagged_mode and not self._trainer.get_model_attribute_full(model=model_name, attribute='val_in_fit', func=max):
```
Comment (Innixma) [inline tabular/src/autogluon/tabular/predictor/predictor.py:898]: What scenario does this fix?

File: tabular/src/autogluon/tabular/trainer/auto_trainer.py [hunk 2]
```diff
@@ -68,7 +71,108 @@ def fit(self, X, y, hyperparameters, X_val=None, y_val=None, X_unlabeled=None, h
...
+            aux_models += self.generate_quantile_aggregator(X=X_stack_preds, y=y,
+                                                            level=level, base_model_names=base_model_names, k_fold=0, n_repeats=1,
+                                                            stack_name=stack_name, time_limit=time_limit, name_suffix=name_suffix,
+                                                            get_models_func=get_models_func, check_if_best=check_if_best,
+                                                            excluded_model_types=excluded_model_types,
+                                                            hyperparameter_tune_kwargs=hyperparameter_tune_kwargs)
+
+        return aux_models
+
+    def generate_quantile_aggregator(self, X, y, level, base_model_names, k_fold=0, n_repeats=1, stack_name=None,
```
Comment (Innixma) [inline tabular/src/autogluon/tabular/trainer/auto_trainer.py:100]: Unit test?

File: tabular/src/autogluon/tabular/predictor/predictor.py [hunk 3]
```diff
@@ -791,12 +791,18 @@ def fit(self,
             'excluded_model_types': excluded_model_types,
             'feature_prune_kwargs': kwargs.get('feature_prune_kwargs', None)
         }
+
+        aux_kwargs = {
+            'ag_args': ag_args,
+            'excluded_model_types': excluded_model_types,
+        }
```
Comment (Innixma) [inline tabular/src/autogluon/tabular/predictor/predictor.py:798]: What are the ramifications of passing `ag_args` to aux models? Is this to enable HPO?

File: tabular/src/autogluon/tabular/models/tabular_nn/torch/tabular_nn_torch.py [hunk 4]
```diff
@@ -578,3 +582,191 @@ def load(cls, path: str, reset_paths=True, verbose=True):
             model._architecture_desc = None
             model.model = torch.load(model.path + model.params_file_name)
         return model
+
+
+class TabularQuantileAggregatorTorchModel(TabularNeuralNetTorchModel):
```
Comment (Innixma) [inline tabular/src/autogluon/tabular/models/tabular_nn/torch/tabular_nn_torch.py:587]: Unit Test?

File: tabular/src/autogluon/tabular/models/tabular_nn/torch/tabular_torch_dataset.py [hunk 5]
```diff
@@ -118,6 +121,14 @@ def __init__(self, processed_array, feature_arraycol_map, feature_type_map, prob
                     labels = labels.astype('long')
                 self.data_list.append(labels.reshape(-1, 1))
 
+        # base prediction data
+        if base_preds is not None:
+            base_preds = np.array(base_preds)
+            self.data_desc.append("base_pred")
+            self.base_pred_index = len(self.data_list)
```
Comment (jwmueller) [inline tabular/src/autogluon/tabular/models/tabular_nn/torch/tabular_torch_dataset.py:128]: I feel like this aggregator-specific code is confusing to have in the general `TabularTorchDataset` which is intended for standard pytorch neural nets trained on the original data, not aggregator networks.  

At the very least it would be good to have a comment clarifying what that this code is specific for aggregator networks.  Even better could be if aggregator networks use a different `aggregator_torch_dataset` object that is inherits from `TabularTorchDataset`.

File: tabular/src/autogluon/tabular/models/tabular_nn/torch/tabular_torch_dataset.py [hunk 6]
```diff
@@ -54,6 +54,7 @@ def __init__(self, processed_array, feature_arraycol_map, feature_type_map, prob
                                             (options: 'vector', 'embed')
             problem_type (str): what prediction task this data is used for.
             labels (pd.Series): list of labels (y) if available
+            base_preds: 2D numpy array. Contains raw data of all base learners' predictions as columns
```
Comment (jwmueller) [inline tabular/src/autogluon/tabular/models/tabular_nn/torch/tabular_torch_dataset.py:57]: If you end up keeping this same `TabularTorchDataset` for the aggregator networks, then should clarify this `base_preds` argument is specifically intended for aggregator networks that ensemble predictions from other models.

Also probably want to clarify what the format of these `base_preds` should be specifically for quantile regression applications (Does it only works for that quantile format or for regression/classification as well? if only for quantile, want to clarify that also).

File: tabular/src/autogluon/tabular/models/tabular_nn/torch/torch_network_modules.py [hunk 7]
```diff
@@ -252,5 +257,58 @@ def predict(self, input_data):
             elif self.problem_type == REGRESSION:
                 predict_data = predict_data.flatten()
             if self.problem_type == BINARY:
-                predict_data = predict_data[:,1]
+                predict_data = predict_data[:, 1]
             return predict_data.data.cpu().numpy()
+
+
+class DeepQuantileAggregator(EmbedNet):
```
Comment (jwmueller) [inline tabular/src/autogluon/tabular/models/tabular_nn/torch/torch_network_modules.py:264]: I'd move to different file

File: tabular/src/autogluon/tabular/models/tabular_nn/torch/tabular_nn_torch.py [hunk 8]
```diff
@@ -578,3 +582,191 @@ def load(cls, path: str, reset_paths=True, verbose=True):
...
+    """
+    PyTorch quantile aggregator based on neural network models that operate on tabular data to aggregate quantile predictions.
```
Comment (jwmueller) [inline tabular/src/autogluon/tabular/models/tabular_nn/torch/tabular_nn_torch.py:589]: add arxiv link for more detailed description?

Comment (jwmueller) [inline tabular/src/autogluon/tabular/models/tabular_nn/torch/tabular_nn_torch.py:589]: add arxiv link for more detailed description?

File: tabular/src/autogluon/tabular/models/tabular_nn/torch/tabular_nn_torch.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jwmueller) [reply tabular/src/autogluon/tabular/models/tabular_nn/torch/tabular_nn_torch.py:587]: I'd also shorten the name to `QuantileAggregatorTorchModel` or `DeepQuantileAggregatorModel`

File: tabular/src/autogluon/tabular/models/tabular_nn/hyperparameters/parameters.py [hunk 9]
```diff
@@ -99,9 +100,9 @@ def get_quantile_hyper_params(framework):
     """ Parameters that currently can be searched during HPO """
     hyper_params = get_hyper_params(framework)
     new_hyper_params = {
-        'gamma': 5.0,  # margin loss weight which helps ensure noncrossing quantile estimates
-        # Options: range(0.1, 10.0)
-        'alpha': 0.01,  # used for smoothing huber pinball loss
+        'gamma': 1.0,  # margin loss weight which helps ensure noncrossing quantile estimates
+        'num_layers': 3,
+        'hidden_size': 128,
```
Comment (gradientsky) [inline tabular/src/autogluon/tabular/models/tabular_nn/hyperparameters/parameters.py:105]: From a practical side I haven't seen MLP performing better with 3 layers vs 2 layers. hidden_size can be larger. What's more important for MLP: 1) the order of the components in the layers, 2) training schedule; 3) normalization.

also why not using different sizes of layers? [200, 100]

**Comment by Innixma** [comment]
Closing due to inactivity.
